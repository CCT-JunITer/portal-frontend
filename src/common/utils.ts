

export const debounce = (fn: (...params: any[]) => any, delay = 1000, immed = false) => {
  let timer: NodeJS.Timeout | null = null;
  return function (this: any, ...args: any[]) {
    if (timer === null && immed) {
      fn.apply(this, args);
    }
    timer && clearTimeout(timer);
    timer = setTimeout(() => { 
      fn.apply(this, args)
      timer = null
    }, delay);
    return timer;
  }
};
