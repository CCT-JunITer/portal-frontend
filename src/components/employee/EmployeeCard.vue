<template>
  <v-card
    class="text-center"
    width="100%"
    v-bind="$attrs">
    <router-link
      :to="{ path: `/main/people/profile/view/${employee.id}`}"
      v-slot="{ navigate, isActive }"
    >
      <div v-ripple :active="isActive" @click="navigate">
        <backdrop class="d-flex align-center justify-center" color="grey lighten-3" height="84%">
          <employee-profile-picture :employee="employee" class="mt-5" size="72"></employee-profile-picture>
        </backdrop>
        <div class="pa-2">
          <div class="text-subtitle-1 text--primary d-flex align-content-start justify-center">
            {{ employee.full_name || 'Kein Name' }}
            <v-icon v-if="hasBirthday" color="cctOrange" size="20">mdi-cake-variant</v-icon>
          </div>
          <div v-if="!isAlumni">
            <div class="text-subtitle-2 mb-2">
              {{ employee.ressort || 'Kein Ressort' }}

              <group-icon
                v-for="group in groups"
                :key="group.id"
                :group="group"
              >
              </group-icon> 
            </div>
            <v-chip small color="cctBlue" outlined dark label>{{ employee.memberstatus }}</v-chip>
          </div>
          <div v-else>
            <div class="text-body-2 mb-2">
              Alumni seit <b>{{ exitFormatted }}</b>
            </div>
            <v-chip small color="cctOrange" outlined dark label>{{ yearsIn }} im Verein</v-chip>
          </div>
        </div>
      </div>
    </router-link>
    <v-divider></v-divider>
    <v-card-actions class="d-flex justify-space-between py-2 px-2 flex-wrap">
      <v-btn
        text
        small
        :href="`mailto:${employee.email}`"
        color="grey"
        class="center-employee-text">
        <v-icon left small>mdi-message</v-icon>
        Nachricht
      </v-btn>
      <v-btn
        text
        small
        :href="`tel:${employee.phonenumber}`"
        color="grey"
        class="center-employee-text">
        <v-icon left small>mdi-phone</v-icon>
        Anrufen
      </v-btn>
    </v-card-actions>
  </v-card>

</template>

<script lang="ts">
import { Component, Vue, Prop } from 'vue-property-decorator';
import {IUserProfile} from '@/interfaces';
import EmployeeProfilePicture from '@/components/employee/EmployeeProfilePicture.vue';
import Backdrop from '@/components/Backdrop.vue';
import GroupIcon from './GroupIcon.vue';
import { format, formatDistance } from 'date-fns';
import { de } from 'date-fns/locale';

@Component({
  components: { Backdrop, EmployeeProfilePicture, GroupIcon }
})
export default class EmployeeCard extends Vue {
  @Prop()
  private employee!: IUserProfile;

  get hasBirthday() {
    if(!this.employee) {
      return false;
    }
    return this.$common.isTodayBirthday(this.employee.birthdate);
  }

  get isAlumni() {
    return this.employee.is_alumni;
  }

  get exitFormatted() {
    return format(new Date(this.employee.exitdate), 'MMMM yyyy', { locale: de });
  }

  get yearsIn() {
    return formatDistance(new Date(this.employee.exitdate), new Date(this.employee.entrydate), { locale: de })
  }


  get groups() {
    return this.employee.active_groups;
  }
}
</script>

