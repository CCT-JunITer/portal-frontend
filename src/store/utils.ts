import { AxiosError, AxiosResponse } from 'axios';
import { ActionContext } from 'vuex';
import { dispatchCheckApiError } from './main/actions';
import { commitAddNotification, commitRemoveNotification } from './main/mutations';
import { AppNotification, MainState } from './main/state';
import { State } from './state';

export const replace = <T extends {id: number}>(array: T[] | null, payload: T): T[] => {
  if (!array) {
    return [payload];
  }
  const replaced = array.map(item => item.id !== payload.id ? item : payload)
  if(!replaced.find(item => item.id === payload.id)) {
    replaced.push(payload);
  }
  return replaced;
}

export const apiCallNotify = async <T>(context: ActionContext<unknown, State>, apicall: (token: string) => Promise<AxiosResponse<T>>, options: {successText?: string | null; loadingText?: string | null, errorCallback?: (e: AxiosError) => Promise<boolean>} = {}) => {
  const { loadingText = 'Lade', successText = 'Erfolgreich gespeichert' } = options;
  let loadingNotification: AppNotification | null = null;
  if (loadingText) {
    loadingNotification = { content: loadingText, showProgress: true, color: 'cctBlue' };
  }
  try {
    loadingNotification && commitAddNotification(context, loadingNotification);
    const response = await apicall(context.rootState.main.token);
    loadingNotification && commitRemoveNotification(context, loadingNotification);
    if (successText) {
      commitAddNotification(context, { content: successText, color: 'success' });
    }
    return response;
  } catch (e: unknown) {
    const error = e as AxiosError<{detail: any}>;
    if (!options.errorCallback || !await options.errorCallback(error)) {
      loadingNotification && commitRemoveNotification(context, loadingNotification);
      commitAddNotification(context, { content: `Fehler ${error.response?.status}: ${error.response?.data?.detail || error.message}`, color: 'error' });
      await dispatchCheckApiError(context, error);
    }
    throw error;
  }
};


export const apiCall = async <T>(context: ActionContext<unknown, State>, apicall: (token: string) => Promise<AxiosResponse<T>>, options: { errorCallback?: (e: AxiosError) => Promise<boolean> } = {}) => {
  try {
    const response = await apicall(context.rootState.main.token);
    return response;
  } catch (e: any) {
    const error = e as AxiosError<{detail: any}>;
    if (!options.errorCallback || !await options.errorCallback(error)) {
      await dispatchCheckApiError(context, error);
    }
    throw error;
  }
};
