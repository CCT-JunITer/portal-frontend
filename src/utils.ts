const getLocal = (key: string) => () => localStorage.getItem(key);
const saveLocal = (key: string) => (value: string) => localStorage.setItem(key, value);
const removeLocal = (key: string) => () => localStorage.removeItem(key);

export const getLocalToken = getLocal('token');
export const saveLocalToken = saveLocal('token');
export const removeLocalToken = removeLocal('token');

export const getLocalUserStatus = getLocal('userstate');
export const saveLocalUserStatus = saveLocal('userstate');
export const removeLocalUserStatus = removeLocal('userstate');

export function dataURItoBlob(dataURI) {
  // convert base64/URLEncoded data component to raw binary data held in a string
  let byteString;
  if (dataURI.split(',')[0].indexOf('base64') >= 0)
    byteString = atob(dataURI.split(',')[1]);
  else
    byteString = unescape(dataURI.split(',')[1]);

  // separate out the mime component
  const mimeString = dataURI.split(',')[0].split(':')[1].split(';')[0];

  // write the bytes of the string to a typed array
  const ia = new Uint8Array(byteString.length);
  for (let i = 0; i < byteString.length; i++) {
    ia[i] = byteString.charCodeAt(i);
  }

  return new Blob([ia], {type:mimeString});
}
