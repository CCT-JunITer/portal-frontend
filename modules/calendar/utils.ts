import { State } from '@/store/state';
import { ActionContext } from 'vuex';
import { AxiosError } from 'axios';
import { apiCall as apiCallMain, apiCallNotify as apiCallNotifyMain } from '@/store/utils';
import { dispatchActionAuthenticateNextcloud } from './store/actions';


const errorCallback = (context: ActionContext<any, State>) => async (error: AxiosError<any>) => {
  if (error.response?.data?.detail?.type == 'nextcloud authentication error') {
    await dispatchActionAuthenticateNextcloud(context);
    return true;
  }
  return false;
}

type apiCallNotifyType<T> = typeof apiCallNotifyMain<T>;
export const apiCallNotify = async <T>(...args: Parameters<apiCallNotifyType<T>>): ReturnType<apiCallNotifyType<T>> => {
  const [context, apicall, options = {}] = args;
  return await apiCallNotifyMain(context, apicall, {...options, errorCallback: errorCallback(context)});
}


type apiCallType<T> = typeof apiCallMain<T>;
export const apiCall = async <T>(...args: Parameters<apiCallType<T>>): ReturnType<apiCallType<T>> => {
  const [context, apicall, options = {}] = args;
  return await apiCallMain(context, apicall, {...options, errorCallback: errorCallback(context)});
}