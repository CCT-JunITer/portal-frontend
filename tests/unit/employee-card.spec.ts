import { bootstrap, getTestUser} from '@tests/utils';
import EmployeeCard from '@/components/employee/EmployeeCard.vue'

describe('EmployeeCard.vue', () => {
  const { customMount } = bootstrap();

  it('should match snapshot', () => {
    
    const wrapper = customMount(EmployeeCard, {
      propsData: { 
        employee: getTestUser()
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})