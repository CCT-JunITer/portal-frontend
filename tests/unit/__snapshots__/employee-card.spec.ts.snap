// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeCard.vue should match snapshot 1`] = `
<div class="text-center v-card v-sheet theme--light" style="width: 100%;">
  <div>
    <div class="root d-flex align-center justify-center">
      <div class="grey lighten-4 backdrop" style="height: 84%;"></div>
      <div class="v-avatar avatar mt-5 grey lighten-2" style="height: 72px; min-width: 72px; width: 72px;"></div>
    </div>
    <div class="pa-2">
      <div class="text-subtitle-1 text--primary d-flex align-content-start justify-center">
        Test
        <!---->
      </div>
      <div class="text-subtitle-2 mb-2">Juniter</div> <span class="v-chip v-chip--label v-chip--outlined theme--dark v-size--small cctBlue cctBlue--text"><span class="v-chip__content">Consultant</span></span>
    </div>
  </div>
  <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
  <div class="v-card__actions d-flex justify-space-between py-2 px-2 flex-wrap"><a href="mailto:test@cct-ev.de" class="center-employee-text v-btn v-btn--flat v-btn--text theme--light v-size--small grey--text"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-message theme--light" style="font-size: 16px;"></i>
      Nachricht
    </span></a> <a href="tel:+49 1231234123" class="center-employee-text v-btn v-btn--flat v-btn--text theme--light v-size--small grey--text"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate v-icon--left mdi mdi-phone theme--light" style="font-size: 16px;"></i>
      Anrufen
    </span></a></div>
</div>
`;
